File: Data Structures & Algorithms/maximum-product-subarray/submission-2.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int n = nums.size();
        int curmax = 1;
        int curmin = 1;
        int ans = nums[0];
        for(int i=0;i<n;i++){
            int temp = nums[i]*curmax;
            curmax = max(max(curmax*nums[i],curmin*nums[i]),nums[i]);
            curmin = min(min(temp,curmin*nums[i]),nums[i]);
            ans = max(curmax,ans);
        }
        return ans;
    }
};

File: Data Structures & Algorithms/rotate-array/submission-2.java
class Solution {
    public void rotate(int[] nums, int k) {
        int i = 0;
        int mid = nums.length-k%nums.length-1;
        int j = nums.length - 1;
        reverse(nums,i,mid);
        reverse(nums,mid+1,j);
        reverse(nums,i,j);
    }

    public void reverse(int[] nums, int i, int j){
        while(i<j){
            int temp = nums[i];
            nums[i] = nums[j];
            nums[j] = temp;
            i++;
            j--;
        }
    }


}